### PR TITLE
Pedersen commitment library

### PR DESCRIPTION
**Pull request by MartinSchmidt**
Added a first draft of the Pedersen commitment.

### CODE REVIEW

File: src/EnergyOrigin.PedersenCommitment/Commitment.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        var rSum = BigInteger.Zero;
+        foreach (var rEl in r)
+        {
+            rSum += rEl;
+        }
+
+        return new Commitment(BigInteger.ModPow(group.g, m, group.p) * BigInteger.ModPow(group.h, rSum, group.p) % group.p, group);
+    }
+
```
Comment (wisbech) [inline src/EnergyOrigin.PedersenCommitment/Commitment.cs:28]: Is it valuable to add a +(addition) and -(subtraction) operator as well ?
My notes says that they are the same as * / but using mod q.

File: src/EnergyOrigin.PedersenCommitment/Group.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    private static BigInteger GenerateGH(Random r, BigInteger p, int k, BigInteger? g = null)
+    {
+        BigInteger gh;
+        do
+        {
+            var a = r.NextBigInteger(2, p - 1); // Select a (1 < a < p-1);
+            gh = BigInteger.ModPow(a, k, p); // Compute g = a^k mod p (remember p = k*q + 1)
+        }
+        while (gh % p == 1 % p || gh == g); // If g == 1 mod p then goto 1 and try again. Otherwise return g.
```
Comment (wisbech) [inline src/EnergyOrigin.PedersenCommitment/Group.cs:93]: Lovely approach (y)

**Review by wisbech — COMMENTED** [review]
Was it the dividend that flipped the modulo operation to be positive ?

Great work (y)

File: src/ProjectOrigin.PedersenCommitment/Commitment.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+using System.Numerics;
+using System.Runtime.Serialization;
+
+namespace ProjectOrigin.PedersenCommitment;
+
```
Comment (wisbech) [inline src/ProjectOrigin.PedersenCommitment/Commitment.cs:5]: A commented line referring to the project origin registry documentation on [Pedersen Commitment](https://github.com/project-origin/registry/blob/main/doc/pedersen_commitments/pedersen-commitments.md)

**Review by wisbech — APPROVED** [review]
Great work on this Pedersen Commitment skeleton - hattip @MartinSchmidt.

An observations going through the code 

- a lot of return statements have function calls and are harder to read from this practice. Most often the most vital part of the code is wrapped in the return statement. I know it is style, but it is harder to read.

File: src/ProjectOrigin.PedersenCommitment/Commitment.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+public record Commitment
+{
+    public BigInteger C { get; }
+
+    [IgnoreDataMember]
```
[the remark this answers is not in the document]
Comment (MartinSchmidt) [reply src/ProjectOrigin.PedersenCommitment/Commitment.cs:10]: It is if one wants to serialize a Commitment info json or etc. then one wouldn't want the group included in each message, so only the actual C would be serialized.

**Pull request merged**